src/notebooks/controllers/preferredKernelConnectionService.ts

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { CancellationToken, NotebookDocument } from 'vscode';
+import { ContributedKernelFinderKind, IContributedKernelFinder } from '../../kernels/internalTypes';
+import { PreferredRemoteKernelIdProvider } from '../../kernels/jupyter/preferredRemoteKernelIdProvider';
+import {
+    IKernelFinder,
+    KernelConnectionMetadata,
+    LiveRemoteKernelConnectionMetadata,
+    LocalKernelSpecConnectionMetadata,
+    PythonKernelConnectionMetadata,
+    RemoteKernelConnectionMetadata,
+    RemoteKernelSpecConnectionMetadata
+} from '../../kernels/types';
+import { disposeAllDisposables } from '../../platform/common/helpers';
+import { IDisposable } from '../../platform/common/types';
+import { getNotebookMetadata, translateKernelLanguageToMonaco } from '../../platform/common/utils';
+import { IInterpreterService } from '../../platform/interpreter/contracts';
+import { ServiceContainer } from '../../platform/ioc/container';
+import { getLanguageOfNotebookDocument } from '../languages/helpers';
+
+/**
```

Comment:
I wonder if we can separate this change from "Create Python Env" to give us smaller PRs to review.

Reply:
Thats how it started, then thought this would be small enough
Sure thing will do.